casper/src/test/resources/RegistryTest.rho

```diff
@@ -0,0 +1,77 @@
+//requires RhoSpec
+new
+  RhoSpecCh,
+  rl(`rho:registry:lookup`),
+  rs(`rho:registry:insertSigned:secp256k1`),
+  rr(`rho:registry:insertArbitrary`),
+  sign(`rho:test:crypto:secp256k1Sign`),
+  secpVerify(`rho:crypto:secp256k1Verify`),
+  blake2b256(`rho:crypto:blake2b256Hash`),
+  test_secp,
+  test_insertSignedAndLookup,
+  test_insertArbitraryAndLookup
+in {
+  rl!(`rho:id:zphjgsfy13h1k85isc8rtwtgt3t9zzt5pjd5ihykfmyapfc4wt3x5h`, *RhoSpecCh) |
+  for(@(_, RhoSpec) <- RhoSpecCh) {
+    @RhoSpec!("testSuite",
+      [
+        ("Verifying a signature should work", *test_secp),
+        ("InsertSigned and lookup should work", *test_insertSignedAndLookup),
+        ("InsertArbitrary and lookup should work", *test_insertArbitraryAndLookup)
+      ])
+  } |
+
+  contract test_secp(rhoSpec, _, ackCh) = {
+    match ("04cc94ab15247b0db1b8219388218eb7461ac74b3d88d4da8c1816fac8e258f5e0f5db9db6bb82b0cc066589dfe77a0d7449db295dab248fb93855ba91813154a9".hexToBytes(),
+           "4d24bf7b37a4b3dc417b8e0e03a0f232b31a535f4880a73575ac855acae49640".hexToBytes()) {
+      (pk, sk) => {
+        new signCh, verifyCh, retCh, hashCh in {
+          blake2b256!("foo", *hashCh) |
+          for (@hash <- hashCh) {
+            sign!(hash, sk, *signCh) |
+            for (@sig <- signCh) {
+              secpVerify!(hash, sig, pk, *verifyCh) |
+              for (@b <- verifyCh) {
+                retCh!(b) |
+                rhoSpec!("assert", (true, "== <-", *retCh), "Expecting true", *ackCh)
+              }
+            }
+          }
+        }
+      }
+    }
+  } |
+
+  contract test_insertSignedAndLookup(rhoSpec, _, ackCh) = {
+    match (1,
+           "foo",
+           "04cc94ab15247b0db1b8219388218eb7461ac74b3d88d4da8c1816fac8e258f5e0f5db9db6bb82b0cc066589dfe77a0d7449db295dab248fb93855ba91813154a9".hexToBytes(),
+           "4d24bf7b37a4b3dc417b8e0e03a0f232b31a535f4880a73575ac855acae49640".hexToBytes()) {
+      (nonce, data, pk, sk) => {
+        new uriCh, signCh, valueCh, hashCh in {
+          blake2b256!((nonce, data), *hashCh) |
```

Comment:
we need additional tests:
 - inserted nonce is bigger than the present one on re-insertion
 - make sure we don't integer-overflow the nonce (nonce = MAX_LONG means no update of the entry is possible)
 - lookup of inexistent uri returns Nil (or whatever the old registry does)
 - insertSigned with wrong signature returns Nil (or whatever the old registry does)

Reply:
I'm not sure if the first two tests make sense. Are you assuming here that insertion increases the nonce? It's the user who has to do this.

Reply:
How did the old registry behave?

I think we rely on the first two points for the security of system contracts. Their private keys used for deploy are hardcoded in our sources after all...

Reply:
If you try to insert with a nonce bigger than MAX_LONG you'll get a `java.lang.NumberFormatException: For input string: "9223372036854775808"` from Rholang. Same should be true for the old registry.

Reply:
We should rather return Nil in the new registry.

Anyway, the more important question is: does the old registry enforces monotonic nonce?

If not, what are the consequences to system process security?

Reply:
> We should rather return Nil in the new registry.

Sounds good.
> does the old registry enforces monotonic nonce?

No, it enforces a strictly monotonic nonce. Trying to insert with a nonce thats `<=` the old nonce failed and fails with `Nil`.

Reply:
I've read through the above and the code again - I've mistaken the properties I want guaranteed with the tests. You're right, those two lines make no sense :)